BENCH: sparse.csgraph.dijkstra: add benchmark

Kai-Striega opened the pull request:
gh-20717 was the first of multiple planned improvements to SciPy. We should track performance changes appropriately for any future work.

<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
None, See gh-20717 for context

#### What does this implement/fix?
This benchmark tracks the performance of the changes in gh-20717. As there are more changes that may be coming, we should have the infrastructure set up to track performance. 

#### Additional information
This benchmark has been adapted from @czgdp1807's work from this [comment](https://github.com/scipy/scipy/pull/20717#issuecomment-2562795171)

Kai-Striega committed:
Add benchmark discussed in gh-20717

gh-20717 was the first of multiple planned improvements to SciPy. We should
track perf changes appropriately for this.

Kai-Striega committed:
STY: Fix line length

We need the line length to be <88 chars

[lint only]

j-bowhay commented on benchmarks/benchmarks/sparse_csgraph_dijkstra.py:
```diff
@@ -40,3 +40,24 @@ def time_dijkstra_multi(self, n, min_only, format):
+    param_names = ["n", "density"]
+
+    def setup(self, n, density):
+       G = scipy.sparse.rand(n, n, density, format='csr') * 100
```
Nit, we should use the new `scipy.sparse.random_array` interface

Kai-Striega replied:
Good pickup, I'll edit that today.

I was following the example from the issue and didn't think.

Kai-Striega committed:
BENCH/ENH: sparse.csgraph.dijkstra: Use new random

This commit uses the ``sparse.random_array`` interface rather than the
 (now outdated) ``sparse.rand`` interface.

 [skip actions] [skip cirrus]

Kai-Striega commented on benchmarks/benchmarks/sparse_csgraph_dijkstra.py:
```diff
@@ -55,8 +57,14 @@ class DijkstraDensity(Benchmark):
+            shape=(n, n),
+            density=density,
+            format='csr',
+            rng=rng,
+        ) * 100
+        self.graph = G.astype(np.uint32)
```
This could be simplified by using the ``data_sampler`` argument from [random_array](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.random_array.html), however I haven't figured out how yet. I'll update this later if people are interested.

j-bowhay replied:
I have no issue with the current approach, using `data_sampler` might be overkill

dschult replied:
The advantage of using `data_sampler` directly gets pretty small between `n=100` and `n=1000`.
But FYI one way to use data_sampler would be:
```python
        G = scipy.sparse.random_array(
            shape=(n, n),
            density=density,
            format='csr',
            rng=rng,
            data_sampler=lambda size : rng.integers(100, size=size, dtype=np.uint32)
        )
```

j-bowhay commented:
@Kai-Striega would you mind marking the `n=1000` case as slow?

Kai-Striega commented:
> @Kai-Striega would you mind marking the `n=1000` case as slow to avoid it running in ci?

Sure thing. I'll also add the data_sampler as I think it looks neater

Kai-Striega committed:
BENCH: sparse.csgraph.dijkstra: Skip n=1000

Kai-Striega committed:
BENCH: sparse.csgraph.dijkstra: use data_sampler

[skip actions] [skip cirrus]

j-bowhay commented on benchmarks/benchmarks/sparse_csgraph_dijkstra.py:
```diff
@@ -1,11 +1,13 @@
 """benchmarks for the scipy.sparse.csgraph module"""
+from functools import partial
+
```
```suggestion
```

j-bowhay committed:
Removed unused import

[skip ci]

j-bowhay reviewed (approved):
Thanks @Kai-Striega !
